Fix searching the main output window which was not accessible to screen readers

hp77-creator opened the pull request:
/claim #7185



Added key shortcut to navigate found results
Created an Announcer object to read the results.

hp77-creator committed:
Fix talkback for search

ironcross32 commented:
I'll echo my sentiments from Discord here.

This is progress, but it doesn't put the focus in the output window. I was initially confused, because I thought it trapped me in the output window, but it never left the search bar, just announced. If I'm searching, it's because I need to copy something out most likely, so the cursor needs to jump to the start of the search term and focus needs to be put into the output window.

hp77-creator committed:
Add caret and ability to move around result

hp77-creator commented:
@ironcross32 please verify the changes. Thanks

ironcross32 commented:
This is better. Ideally, it would be even more so if the cursor jumps to the start of the search term, but how it is now is OK if that's too nit-picky.

Example, if I search for, "cuirass,", it should jump to thee start of the word.

hp77-creator commented:
I saw this issue too, code wise I am always going to the place wherever the text is found, but sometimes it is highlighting the first line, sometimes the start and sometimes at the end, I can see some formatting issue and position not matching with the result. @vadi2 what do you suggest?

vadi2 commented:
I see what @ironcross32 is asking for but I'm not sure what you @hp77-creator mean.

hp77-creator commented:
@vadi2 
In my changes, if you will see, I am using the position found during search, I am placing caret there and announcing the text. Now on screen, this is sometimes, matching correctly with the start of our exact word, sometimes end or maybe sometimes at the start of the line. I mean to say that searchX or coordinates aren't coming properly hence this inconsistency. I am asking for your suggestion, if we should look into this search changes or is it fine the way it is now?

hp77-creator commented:
@ironcross32 should the announcer also announce from the start of the result or from the start of the line?

ironcross32 commented:
right now, that seems to be overridden by the announcement about moving to the past content.

hp77-creator commented:
@ironcross32 so you are saying that, announcement is not working the way it should?
what is your expectation? I will try to implement that, I have been able to implement the cursor starting from search result.

vadi2 commented on src/TConsole.cpp:
```diff
@@ -546,6 +552,9 @@ TConsole::TConsole(Host* pH, const QString& name, const ConsoleType type, QWidge
     }
 
+    // Create Announcer for screen reader support
+    new Announcer(this);
```
You can use the existing `mudlet::announce()` to announce (if needed)

hp77-creator replied:
removed

vadi2 commented on src/TConsole.cpp:
```diff
@@ -1840,9 +1849,24 @@ void TConsole::slot_searchBufferUp()
+            
+            // Announce match to screen readers
+            if (auto* announcer = findChild<Announcer*>()) {
+                announcer->announce(buffer.lineBuffer[searchY]);
```
I am not sure this is needed given what @ironcross32 is saying. Is Voiceover already not speaking things from the setCaretPosition changes?

hp77-creator replied:
I had added these to read the whole line, let me see, also, I have not pushed the caret update change yet. will do.

hp77-creator replied:
@vadi2 this will be needed, and I will move them to the place where highlighting is being done, that will be helpful, although, I am observing that voiceover starts dictating words in the active window and output window content is not read sometimes

ironcross32 commented:
Yes, what I'm saying is that the announcement of the search result no longer happens as it did in the first build because instead, the screen reader speaks the fact that the cursor has been moved to the past content.

hp77-creator commented:
got it, @ironcross32, thanks for your swift response. let me fix it and get back to you.

hp77-creator committed:
refactor and move announce to search

hp77-creator commented:
@ironcross32 
i have tried to move announcement to the line search result is found, During my testing I observed that other active window was given more precedence, you might have to wait a bit for the output window to take focus. Let me know if this is okay.

cc: @vadi2

ironcross32 commented:
yes, this is better. The only other thing I might recommend doing is making an option in the accessibility tab of preferences to enable the shortcut, and make it enabled when a user types `mudlet access on`. There are folks who might not appreciate losing an F-key because they are good for keybinds.

hp77-creator commented:
where do we enter the command? @ironcross32 @vadi2

vadi2 commented:
It would be in this menu: 

![image](https://github.com/user-attachments/assets/bba63b05-643d-4707-b47a-ec3e6d43318d)

Add it to the [setConfig function](https://github.com/Mudlet/Mudlet/blob/development/src/TLuaInterpreter.cpp#L7096) to allow programmatic access, and then you can add to the mudlet access on alias, just like we have an option for setting the [switch key](https://github.com/Mudlet/Mudlet/blob/development/src/mudlet-lua/lua/enable-accessibility/enable-accessibility.xml#L17-L18).

hp77-creator commented:
Yes, I understood that part, I have also included it. I want to know where do we enter `mudlet access on` command?

vadi2 commented:
In the input line, it's a default alias that is installed on all new profiles: https://wiki.mudlet.org/w/Manual:Screen_Readers

hp77-creator commented:
@vadi2 I have a suggestion, can we merge this PR? what @ironcross32 has suggested can be taken up as an enhancement since the original issue didn't have this requirement and currently for some reason, I am unable to get the dialog box where we enter the commands, what do you think? 
In my opinion, the goal of the issue was achieved but adding the capability to toggle off extra shortcut was not in the scope and it is also not working in my system.

vadi2 commented:
It would be hard to accept it if it adds a feature but breaks the keybinding for others :sweat_smile: 

What is it that you are having trouble with?

hp77-creator commented:
For some reason package for enabling accessibility settings is not coming up on my install

You can check this : https://discord.com/channels/283581582550237184/283582439002210305/1321133638783602789

and the replies

vadi2 commented:
The prompt is in the black bar at the bottom of the screen. It doesn't always get an outline for some reason, but you can click on it and type in it!

hp77-creator commented:
> The prompt is in the black bar at the bottom of the screen. It doesn't always get an outline for some reason, but you can click on it and type in it!

I have tried entering the command but it doesn't detect the command

hp77-creator commented:
@vadi2 I have added some changes to xml file for enable-accessibility, when I build mudlet locally and try to test, I don't see the changes that I add in the xml in the package manager, do we have to load it differently?


what I mean is I have added my changes in the xml file like following:
<img width="960" alt="image" src="https://github.com/user-attachments/assets/3da1f3bc-0ffa-45c8-a480-1f1234c370b7" />


but when I load it in the triggers and alias, I see only this:
<img width="960" alt="image" src="https://github.com/user-attachments/assets/9c60b35b-48aa-4a84-a941-7b6004a946b6" />

atari2600tim commented:
> when I build mudlet locally and try to test, I don't see the changes that I add in the xml in the package manager, do we have to load it differently?

Are you trying this on a new profile or an existing profile?  The default packages for a profile get set up when creating the profile but then in the next session it will just load the saved data.  So you should see the changes if you create a new profile.

Since I see `3Kingdoms` in the screenshot, another thing to keep in mind is that deleting a profile in the connection window will be treated special when it is one of the default games.  It will add the deleted profile into a list of games that the person has manually deleted held in Windows registry, so it will avoid recreating it.  So if you want to start fresh the simplest way is to go into the profiles folder and delete the whole folder there.  Or you can go into the registry and empty that list and then next session it will set it up.

vadi2 commented:
@hp77-creator we do need to load it differently! Uninstall the package, install it as a module instead, and enable **Sync**. This will then update the XML on disk whenever you save the profile.

hp77-creator commented:
@vadi2 is there any command like mudlet accessibility off? I don't see, the command for that in the package

ZookaOnGit commented:
> is there any command like mudlet accessibility off?

nope, no command for off

hp77-creator committed:
Add F3 option to enable F3 key for search

hp77-creator committed:
Code refactor

hp77-creator commented:
@vadi2 you can review
@ironcross32 you can test, I have enabled the shortcut with `mudlet access on`, by default it will be disabled.

ironcross32 commented:
I have a very odd test result. First of all, in a profile I already had, the function is off by default, as it should be. I enable it, start using it, all is well. I then go to test, and I pick a profile out of random, which happens to be Medievia. I don't have a character there, have never played there, so I load it in offline mode, do `mudlet access on` and `mudlet access reader` which I do for every new profile. The F3 shortcut is not working. I connect to the profile and figure I can at least search the welcome screen, but still, the shortcut doesn't work.

I then go to the accessibility options, see that it is checked, so I decide to switch it off, then back on to see if that helps. Nope. I go back in, it's off, I don't know why it's off, so I turn it back on and try again, still no.

I then create a brand new profile that points to a MUD that's being hosted locally on another machine in my home. Again, `mudlet access on` and `mudlet access reader`. This time, no issue at all with the F3 shortcut working.

I don't know if there's something different about profiles that ship with Mudlet or what.

hp77-creator commented:
> I have a very odd test result. First of all, in a profile I already had, the function is off by default, as it should be. I enable it, start using it, all is well. I then go to test, and I pick a profile out of random, which happens to be Medievia. I don't have a character there, have never played there, so I load it in offline mode, do `mudlet access on` and `mudlet access reader` which I do for every new profile. The F3 shortcut is not working. I connect to the profile and figure I can at least search the welcome screen, but still, the shortcut doesn't work.

I had tested on my system after clearing out the existing data, it had worked then

> I then go to the accessibility options, see that it is checked, so I decide to switch it off, then back on to see if that helps. Nope. I go back in, it's off, I don't know why it's off, so I turn it back on and try again, still no.

>I don't know if there's something different about profiles that ship with Mudlet or what.


Can you reload the profiles or check the lua scripts, I am suspecting that the shortcut command was not triggered for existing profiles, this was the case for me as well, I had to remove enable-accessibility and then install it again from the source code, can you try with that?

jmckisson commented:
Medievia may have a key defined already for F3 if it was allowed to install the UI. I am the package maintainer for that and I recall F3 being a shortcut for a spell, not at my desktop atm to verify.

ironcross32 commented:
OK, I tested with another default profile which I never used before, and this time it works. I did find the F3 key within the Medievia UI.

Normally, I'm a fan of being able to override system keybinds with user-created ones. See MUSHclient's Accelerator() function. But in this case, I think because it's specifically for enabling accessibility, it should take top priority.

SlySven commented:
Apropos https://github.com/Mudlet/Mudlet/pull/7579#issuecomment-2566516056 when I can get PR 7428 sorted the problem with deleting predefined MUDs should be cured as a side effect - see: https://github.com/Mudlet/Mudlet/pull/7428#issuecomment-2341323868 ...

hp77-creator committed:
Merge branch 'development' into fix_talkback_search

hp77-creator commented:
@vadi2 can you review this?

vadi2 reviewed (commented):
This seems to break searching, here is what I found:

I echoed `removed` 6 times, went to the search bar, typed 'removed' and then press enter 3 times. In 4.19.1, it keeps going up a result - in your PR, it only goes up once and stops.

Additionally, when I use click the button to go up manually, the first letter of 'removed' gets highlighted oddly, and there is a screen flicker every time I search.

[Screencast from 2025-01-05 16-22-38.webm](https://github.com/user-attachments/assets/686ef571-ef5d-42e5-8184-3426970d871b)

Please test against 4.19.1 to see how search compares - it should work exactly the same way, with the only improvement being the addition of accessibility features.

hp77-creator commented:
screen flicker happens because it has to put the marker at the first letter so that the reader can find that, I can check why `enter` is not working,

vadi2 commented:
The screen flicker does not happen in 4.19.1 though, and it would be an a negative impact user experience, unfortunately.

hp77-creator commented:
it basically re-renders with the marker, i am not sure how can I put a marker without rendering it again, without putting a marker, user can't copy from the reader's position

hp77-creator committed:
Merge branch 'development' into fix_talkback_search

hp77-creator commented:
i can disable the re-render for default, but it needs to happen in accessibility mode to allow the reader feature

hp77-creator committed:
Add condition to only allow F3

hp77-creator commented:
@vadi2 you may check now

vadi2 commented:
The shortcut doesn't seem to be working, pressing F3 nor Shift F3 doesn't move the results up or down - both on Windows and Linux. Can you confirm it's working?

hp77-creator commented:
did u try after enabling it using `mudlet access on`? @vadi2

vadi2 commented:
Yep, and it's enabled in the menu

hp77-creator committed:
Merge branch 'development' into fix_talkback_search

hp77-creator commented:
Also the game was not 'Medievia' right?

vadi2 commented:
Yes. Try this - create a new profile, and don't do `mudlet access on` but go to accessibility settings and tick `Enable F3 shortcuts` by hand - it didn't seem to trigger this way.

A second problem, open 3 more profiles and enable F3 shortcuts in all of them and then try searching in one - it seems like it does multiples of searches at once.

Once we get the bugs ironed out, it'll be good to go :star2:

hp77-creator committed:
Merge branch 'development' into fix_talkback_search

hp77-creator committed:
Fix multiple search indexes

hp77-creator commented:
@vadi2 
you can test now, multi jumps by caret are resolved. Also if you don't see the following text: `F3 search functionality enabled ✓` when you do `mudlet access on`, I'd suggest to reload the `enable-accessibility` package and then try again.
Thanks.

vadi2 commented:
The first issue I mentioned still is there:

https://github.com/user-attachments/assets/ad762791-2a36-4908-9215-104ad074325a

I couldn't test the second issue as pressing `F3` seems to say 'no search results found':


https://github.com/user-attachments/assets/6fd387a3-dca4-4678-a5e1-4d28d26773a6

When you fix both issues, please upload a screencast showing the problems are resolved.

hp77-creator commented:
I didn't do any change:
for the first one: https://drive.google.com/file/d/1pTjVBtoLq2FCjFbFek8cwNCKEQQZywRd/view?usp=sharing

Can you check if the `enable-accessibility` is properly loaded for you?

hp77-creator commented:
for the second one, pressing F3 is pressing `down key` in search 
if you press `shift F3` it will take you to the above results, if you reach the end of the results, it will print `No Results found`

vadi2 committed:
Merge branch 'development' into fix_talkback_search
